### PR TITLE
SAK-45537 - Access: Ability to consume time-based tokens for secured files

### PR DESCRIPTION
**Pull request by mpellicer**
https://jira.sakaiproject.org/browse/SAK-45537

### CODE REVIEW

File: access/access-impl/impl/src/java/org/sakaiproject/access/tool/AccessServlet.java [hunk 1]
```diff
@@ -346,6 +357,55 @@ public SecurityAdvice isAllowed(String userId, String function, String reference
...
+								return;
+							}
+						} catch (Exception e) {
+							log.error("Error serving a secured resource {}", e.getMessage());
+						} finally {
+							securityService.popAdvisor(securityAdvisor);
+						}
+					}
+				} catch (Exception e) {
+					// Do not reveal failed requests.
```
Comment (ottenhoff) [inline access/access-impl/impl/src/java/org/sakaiproject/access/tool/AccessServlet.java:405]: how about a debug log for people testing out?

log.debug("error while decrypting secure token", e)

Comment (mpellicer) [reply access/access-impl/impl/src/java/org/sakaiproject/access/tool/AccessServlet.java:405]: Thanks @ottenhoff, I've included it.

**Comment by mpellicer** [comment]
@ottenhoff @ern I still need to work on this, it's not finished since we might want to support more question types and answers, but I feel it sounds very promising.

Using a converter makes it kinda low risk at this moment but I'd like to make some improvements.

Tokens are generated for ONE minute for testing purposes so you can take the URL and test after one minute, it will return 403 after the minute.

T&Q requires a lot of testing so this is just a proof of concept, but is testable locally.

**Comment by mpellicer** [comment]
There you go, this is a recording showing the behavior for the instructor and the student.

Current Sakai behavior, without secured mode:
not_secured_file

After setting this patch, with secured mode:
secured_file

**Comment by mpellicer** [comment]
Okay, I've implemented more changes and now it supports almost all kind of questions sensible to attachments.

I made several tests using multiple sessions and browsers, the images are displayed only for the student at the moment of the exam.

**Review by ern — COMMENTED** [review]
Nice work Miguel

**Review by ern — APPROVED** [review]

File: content/content-bundles/resources/content.properties [hunk 2]
```diff
@@ -665,3 +665,5 @@ createUrls.requiredfield.name=Website Name is a required field. Please provide a
 
 access.hidden=(Hidden)
 resource.is.hidden=This resource/collection is hidden
+
+resource.is.secured.access=The secured access is enabled for this file, it can be accessed under special conditions like an assessment.
```
Comment (ern) [inline content/content-bundles/resources/content.properties:669]: This resource is secured and can only be accessed under specific conditions.

File: content/content-bundles/resources/types.properties [hunk 3]
```diff
@@ -14,6 +14,7 @@ access.groups8		 = Resources in the ''{0}'' folder can only be viewed by the gro
 access.groups9		 = Resources in the ''{0}'' folder can only be viewed by the groups <strong>''{1}'', ''{2}'', ''{3}'', ''{4}'', ''{5}'', ''{6}'', ''{7}'', ''{8}'' and ''{9}''</strong>. You can further restrict this file by group if you wish.
 access.groupsX		 = Resources in the ''{0}'' folder can only be viewed by the groups <strong>''{1}'', ''{2}'', ''{3}'', ''{4}'', ''{5}'', ''{6}'', ''{7}'', ''{8}'', ''{9}'' and others</strong>. You can further restrict this file by group if you wish.
 access.inherit		 = Leave access settings as for the folder.
+access.secured = Allow secured access to this file. This file can be accessed by students in special contexts like an assessment attempt.
```
Comment (ern) [inline content/content-bundles/resources/types.properties:17]: Secure access to this resource only restricting access in specific conditions.

Comment (mpellicer) [reply content/content-bundles/resources/types.properties]: Done

File: content/content-bundles/resources/content.properties (hunk 2 above)
Comment (mpellicer) [reply content/content-bundles/resources/content.properties]: Done

**Pull request merged**